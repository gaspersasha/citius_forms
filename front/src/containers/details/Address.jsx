import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import { Typography, InputText, InputPostcode } from '~components';
import { fieldShape } from './Details';
import s from './styles/details.module.sass';

const Address = ({ names, fields, handleChange }) => (
  <>
    <Typography className={cn(s.subheading, s.subheadMargin)}>
      3. Address details
    </Typography>
    <InputPostcode
      id={names.postcode}
      name={names.postcode}
      value={fields.postcode.value}
      status={fields.postcode.status}
      onChange={handleChange}
      className={s.input}
      required
    />
    <InputText
      id={names.houseName}
      name={names.houseName}
      label="House name"
      value={fields.houseName.value}
      placeholder="e.g. house name"
      status={fields.houseName.status}
      onChange={handleChange}
      className={s.input}
    />
    <InputText
      id={names.houseNumber}
      name={names.houseNumber}
      label="House number"
      value={fields.houseNumber.value}
      placeholder="e.g. 11"
      status={fields.houseNumber.status}
      onChange={handleChange}
      className={s.input}
    />
    <InputText
      id={names.street}
      name={names.street}
      label="Street"
      value={fields.street.value}
      placeholder="e.g. Rover street"
      status={fields.street.status}
      onChange={handleChange}
      className={s.input}
    />
    <InputText
      id={names.district}
      name={names.district}
      label="District"
      value={fields.district.value}
      placeholder="e.g. Hackney"
      status={fields.district.status}
      onChange={handleChange}
      className={s.input}
    />
    <InputText
      id={names.town}
      name={names.town}
      label="Town"
      value={fields.town.value}
      placeholder="e.g. London"
      status={fields.town.status}
      onChange={handleChange}
      className={s.input}
    />
    <InputText
      id={names.county}
      name={names.county}
      label="Ð¡ounty"
      value={fields.county.value}
      placeholder="e.g. Middlesex"
      status={fields.county.status}
      onChange={handleChange}
      className={s.input}
    />
  </>
);

Address.propTypes = {
  fields: PropTypes.shape({
    postcode: fieldShape,
    houseName: fieldShape,
    houseNumber: fieldShape,
    street: fieldShape,
    district: fieldShape,
    town: fieldShape,
    county: fieldShape,
  }).isRequired,
  names: PropTypes.shape({
    postcode: PropTypes.string,
    houseName: PropTypes.string,
    houseNumber: PropTypes.string,
    street: PropTypes.string,
    district: PropTypes.string,
    town: PropTypes.string,
    county: PropTypes.string,
  }).isRequired,
  handleChange: PropTypes.func.isRequired,
};

export default Address;
