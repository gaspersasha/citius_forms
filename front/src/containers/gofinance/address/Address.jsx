import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import { Button, InputCheckbox } from '~components';
import { INPUT_STATUS, VALIDATION_TYPES } from '~constants';
import {
  validate,
  prepare,
  simplify,
  isValid,
  findMaxSteps,
  scrollTo,
  checkNotEmpty,
  mergeDeep,
} from '~utils';
import { handleGTM, sendGTM } from '../helpers';
import { PostCodeStep, ResidentialStep, YearsStep } from '.';
import { getInitialState } from './innitialStateConfig';
import s from './address.module.sass';
import financeStyles from '../styles/finance.module.sass';

const nextButtonIDs = {
  HOME_ADDRESS: 'ssFinanceHomeAddressSubmit',
  RESEDENTIAL_STATUS: 'ssFinanceResedentialStatusSubmit',
  YEARS_AT_ADDRESS: 'ssFinanceYearsAtAddressSubmit',
};

const Address = ({
  onSubmit,
  sumOfYearsInPrevForms,
  step,
  index,
  address,
  vehicleKeptAtAddress,
  isDataSending,
}) => {
  const [state, setState] = useState(
    getInitialState(address, sumOfYearsInPrevForms, vehicleKeptAtAddress, index)
  );
  const [maxSteps, setMaxSteps] = useState(0);
  const isDataSendingInProgress = isDataSending && step === 2;

  useEffect(() => {
    setMaxSteps(findMaxSteps(state));
  }, []);

  useEffect(() => {
    if (!state.validAtStep) return;
    onSubmit(simplify(state), index, maxSteps);
  }, [state.validAtStep]);

  const handleResidentStatusChange = ({ target: { value } }) => {
    setState((prev) => ({
      ...prev,
      residentialStatus: {
        ...prev.residentialStatus,
        value,
      },
    }));
  };

  const handleResidenceYearsChange = ({ target: { value } }) => {
    const isMonthDisabled = state.sumOfYearsInPrevForms + Number(value) >= 3;

    // if years are set to 3, month are disabled and user could send previously selected value
    setState((prev) => ({
      ...prev,
      residenceMonths: {
        ...prev.residenceMonths,
        value: value === '3' ? '0' : prev.residenceMonths.value,
      },
      residenceYears: {
        ...prev.residenceYears,
        value,
      },
      isMonthDisabled,
    }));
  };

  const handleResidenceMonthsChange = ({ target: { value } }) => {
    setState((prev) => ({
      ...prev,
      residenceMonths: {
        ...prev.residenceMonths,
        value,
      },
    }));
  };

  const changeAddressField = ({ target: { name, value } }) =>
    setState((prev) => ({
      ...prev,
      [name]: {
        ...prev[name],
        status: INPUT_STATUS.DEFAULT,
        value,
      },
    }));

  const changePostcodeField = (e) => {
    const custom = {
      ...e,
      target: { name: e.target.name, value: e.target.value.toUpperCase() },
    };

    return changeAddressField(custom);
  };

  const selectAddressHandler = (items) => {
    // create signle object for all address fields, to update state in one go
    const freshAddress = Object.keys(items)
      .map((name) => ({
        [name]: {
          status: checkNotEmpty(items[name])
            ? INPUT_STATUS.VALID
            : INPUT_STATUS.DEFAULT,
          value: items[name],
          validationType: state[name].validationType || VALIDATION_TYPES.NONE,
          step,
        },
      }))
      .reduce((final, item) => ({ ...final, ...item }), {});

    if (!freshAddress.postCode.value) sendGTM('addressChange');

    setState({ ...state, ...freshAddress });
  };

  const handleConfirm = ({ target: { checked } }) =>
    setState((prev) => ({
      ...prev,
      vehicleKeptAtAddress: {
        ...prev.vehicleKeptAtAddress,
        status: checked ? 'default' : 'invalid',
        value: checked,
      },
    }));

  const handleSubmit = (e) => {
    e.preventDefault();
    e.stopPropagation();

    if (isDataSendingInProgress) return;

    const updated = validate(prepare(state, step));
    const tryAgain = !isValid(updated);

    tryAgain && scrollTo('.error-input-label');

    if (step === 0) sendGTM('addressConfirm');
    handleGTM(updated, ['residentialStatus', 'residenceYears']);

    setState({
      ...state,
      ...updated,
      validAtStep: tryAgain ? state.validAtStep : state.validAtStep + 1,
    });
  };

  const getNextButtonID = () => {
    switch (step) {
      case 0:
        return nextButtonIDs.HOME_ADDRESS;
      case 1:
        return nextButtonIDs.RESEDENTIAL_STATUS;
      case 2:
        return nextButtonIDs.YEARS_AT_ADDRESS;
      default:
        return null;
    }
  };

  const buttonClasses = cn(
    { 'button-forward-loading': isDataSendingInProgress },
    `${financeStyles.confirm}`
  );

  const filtredState = {};

  useCallback(mergeDeep(filtredState, state));

  return (
    <div className={s.formSectionWrapper}>
      <div className={cn(financeStyles.formPreamble, s.addressPreamble)}>
        We need your&nbsp;
        <b>UK address</b>
        &nbsp;details for the past three years.
        <br />
        Add your address from the past 3 years below
      </div>
      {step === 0 && (
        <>
          <PostCodeStep
            changeAddressField={changeAddressField}
            changePostcodeField={changePostcodeField}
            selectAddressHandler={selectAddressHandler}
            postCode={state.postCode}
            houseName={state.houseName}
            houseNumber={state.houseNumber}
            street={state.street}
            district={state.district}
            town={state.town}
            county={state.county}
          />
          {!vehicleKeptAtAddress &&
            index === 0 &&
            state.vehicleKeptAtAddress.status === 'invalid' && (
              <div
                className={cn(
                  {
                    'warning-label-subtitle':
                      state.vehicleKeptAtAddress.status === 'invalid',
                  },
                  'label-sub-title'
                )}
              >
                You must tick this box to show that you will keep the vehicle at
                the registered address
              </div>
            )}
          {!vehicleKeptAtAddress && index === 0 && (
            <InputCheckbox
              name="I confirm the vehicle will be kept at this address"
              label="I confirm the vehicle will be kept at this address"
              checked={!!state.vehicleKeptAtAddress.value}
              onChange={handleConfirm}
              customClasses="inlineFlex"
              data-id="I confirm the vehicle will be kept at this address"
            />
          )}
        </>
      )}
      {step === 1 && (
        <ResidentialStep
          residentialStatus={state.residentialStatus}
          onChange={handleResidentStatusChange}
        />
      )}
      {step === 2 && (
        <YearsStep
          residenceYears={state.residenceYears}
          residenceMonths={state.residenceMonths}
          handleResidenceYearsChange={handleResidenceYearsChange}
          isMonthDisabled={state.isMonthDisabled}
          handleResidenceMonthsChange={handleResidenceMonthsChange}
        />
      )}
      <Button
        type="submit"
        className={buttonClasses}
        onClick={handleSubmit}
        id={getNextButtonID()}
      >
        {isDataSendingInProgress ? 'Loading...' : 'Next'}
      </Button>
    </div>
  );
};

Address.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  sumOfYearsInPrevForms: PropTypes.number.isRequired,
  step: PropTypes.number.isRequired,
  vehicleKeptAtAddress: PropTypes.bool,
  isDataSending: PropTypes.bool.isRequired,
  address: PropTypes.shape({
    postCode: PropTypes.string,
    houseName: PropTypes.string,
    houseNumber: PropTypes.string,
    street: PropTypes.string,
    district: PropTypes.string,
    town: PropTypes.string,
    county: PropTypes.string,
    buildingName: PropTypes.string,
    buildingNumber: PropTypes.string,
    residentialStatus: PropTypes.string,
    residenceYears: PropTypes.string,
    residenceMonths: PropTypes.string,
  }),
  index: PropTypes.number,
};

Address.defaultProps = {
  address: {
    postCode: '',
    houseName: '',
    houseNumber: '',
    street: '',
    district: '',
    town: '',
    county: '',
    buildingName: '',
    buildingNumber: '',
    residentialStatus: '',
    residenceYears: '',
    residenceMonths: '0',
  },
  vehicleKeptAtAddress: false,
  index: 0,
};

export default Address;
