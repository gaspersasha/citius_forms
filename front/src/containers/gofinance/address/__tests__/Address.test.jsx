import React from 'react';

import { endpointFetch, debounce } from '~utils';
import Address from '../Address';

jest.mock('~utils', () => ({
  ...jest.requireActual('../../../../utils'),
  endpointFetch: jest.fn(),
  debounce: jest.fn(),
}));

mockUseUserContext({
  user: { id: -1 },
});

endpointFetch.mockResolvedValue({
  Items: [
    {
      Id: 'GB|RM|A|25046440',
      Type: 'Address',
      Text: 'F Gyllyng Flats',
      Description: 'Falmouth TR11 3EZ',
      Language: 'ENG',
      SubBuilding: '',
      BuildingNumber: '',
      BuildingName: 'F Gyllyng Flats',
      Street: '',
      District: '',
      City: 'Falmouth',
      PostalCode: 'TR11 3EZ',
    },
  ],
});
debounce.mockImplementation((cb) => cb);

describe('<Address />', () => {
  const props = {
    onSubmit: jest.fn(),
    index: 0,
    sumOfYearsInPrevForms: 0,
    step: 0,
    isLoading: false,
    isDataSending: false,
  };

  const wrapper = mount(<Address {...props} />);

  it('renders without errors', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('invalid form can`t be submitted', () => {
    const selector = 'button';

    wrapper.find(selector).simulate('click');
    expect(wrapper.props().onSubmit).toHaveBeenCalledTimes(0);
  });

  it('"loqate address" input responds to changes', () => {
    wrapper.setProps({ ...props, step: 0 });
    const selector = 'input[type="text"]';
    const event = { target: { value: 'qwerty' } };

    wrapper.find(selector).simulate('change', event);
    expect(wrapper.find(selector).props().value).toBe('qwerty');
  });

  it('user got suggested address', () => {
    wrapper.update();
    const results = 'div[data-id="loqate-results"]';

    expect(wrapper.find(results).text()).toBe(
      'F Gyllyng Flats Falmouth TR11 3EZ'
    );
  });

  it('user can select suggested address', async () => {
    const result = 'div[data-id="loqate-results"] ul li';

    wrapper.find(result).simulate('click');
    await endpointFetch;
    await endpointFetch;
    wrapper.update();
    const selectedResult = '[data-id="loqate-results"]';

    expect(wrapper.find(selectedResult).text()).toBe(
      'F Gyllyng FlatsFalmouthTR11 3EZ'
    );
  });

  it('onSubmit shoundnt be called without checkbox selected', () => {
    const selector = 'button';
    const event = { preventDefault: jest.fn(), stopPropagation: jest.fn() };

    wrapper.find(selector).simulate('click', event);
    expect(wrapper.props().onSubmit).toHaveBeenCalledTimes(0);
  });

  it('onSubmit could be called when checkbox selected', () => {
    const submit = 'button';
    const event = { preventDefault: jest.fn(), stopPropagation: jest.fn() };

    wrapper
      .find(
        'input[data-id="I confirm the vehicle will be kept at this address"]'
      )
      .simulate('change', { target: { checked: true } });

    wrapper.find(submit).simulate('click', event);
    expect(wrapper.props().onSubmit).toHaveBeenCalledTimes(1);
  });

  it('user can clear suggested address', () => {
    const input = 'input[type="text"]';
    const clearAddresLink = 'div[data-id="clear-addres-link"]';

    wrapper.find(clearAddresLink).simulate('click');
    wrapper.update();
    expect(wrapper.find(input).props().value).toBe('');
  });

  it('user can switch to manually entering address inputs', () => {
    const link = 'div[data-id="enter-address-manually"]';

    wrapper.find(link).simulate('click');
    const input = 'input[type="text"]';

    expect(wrapper.find(input)).toHaveLength(7);
  });

  it('inputs for manually entering address responds to changes', () => {
    const input = 'input[type="text"]';
    const postCode = 'input[type="text"][name="postCode"]';

    wrapper.find(input).forEach((node) =>
      node.simulate('change', {
        target: { value: 'asdfgh', name: node.props().name },
      })
    );
    wrapper
      .find(input)
      .not(postCode)
      .forEach((node) => expect(node.props().value).toBe('asdfgh'));
    expect(wrapper.find(postCode).props().value).toBe('ASDFGH');
  });

  it('"residential status" input responds to changes', () => {
    wrapper.setProps({ ...props, step: 1 });
    const selector =
      'input[data-id="Housing association"][value="HOUSING_ASSOCIATION"]';

    wrapper.find(selector).simulate('change');
    expect(wrapper.find(selector).props().checked).toEqual(true);
  });

  it('"years at address" input responds to changes', () => {
    wrapper.setProps({ ...props, step: 2 });
    const selector =
      'select[data-id="How many years have you lived at this address?"]';
    const event = { target: { value: 1, name: 'residenceYears' } };

    wrapper.find(selector).simulate('change', event);
    expect(wrapper.find(selector).props().value).toEqual(1);
  });

  it('"months at address" input responds to changes', () => {
    const selector =
      'select[data-id="How many months did you live at this address?"]';
    const event = { target: { value: '2' } };

    wrapper.find(selector).simulate('change', event);
    expect(wrapper.find(selector).props().value).toBe('2');
  });
});
