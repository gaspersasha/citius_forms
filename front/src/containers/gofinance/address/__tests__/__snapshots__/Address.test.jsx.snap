// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Address /> renders without errors 1`] = `
<Address
  address={
    Object {
      "buildingName": "",
      "buildingNumber": "",
      "county": "",
      "district": "",
      "houseName": "",
      "houseNumber": "",
      "postCode": "",
      "residenceMonths": "0",
      "residenceYears": "",
      "residentialStatus": "",
      "street": "",
      "town": "",
    }
  }
  index={0}
  isDataSending={false}
  isLoading={false}
  onSubmit={[MockFunction]}
  step={0}
  sumOfYearsInPrevForms={0}
  vehicleKeptAtAddress={false}
>
  <div
    className="formSectionWrapper"
  >
    <div
      className="formPreamble addressPreamble"
    >
      We need your 
      <b>
        UK address
      </b>
       details for the past three years.
      <br />
      Add your address from the past 3 years below
    </div>
    <PostCodeStep
      changeAddressField={[Function]}
      changePostcodeField={[Function]}
      county={
        Object {
          "status": "default",
          "step": 0,
          "validationType": "none",
          "value": "",
        }
      }
      district={
        Object {
          "status": "default",
          "step": 0,
          "validationType": "none",
          "value": "",
        }
      }
      houseName={
        Object {
          "status": "default",
          "step": 0,
          "validationType": "none",
          "value": "",
        }
      }
      houseNumber={
        Object {
          "status": "default",
          "step": 0,
          "validationType": "none",
          "value": "",
        }
      }
      postCode={
        Object {
          "status": "default",
          "step": 0,
          "validationType": "postCode",
          "value": "",
        }
      }
      selectAddressHandler={[Function]}
      street={
        Object {
          "status": "default",
          "step": 0,
          "validationType": "notEmpty",
          "value": "",
        }
      }
      town={
        Object {
          "status": "default",
          "step": 0,
          "validationType": "notEmpty",
          "value": "",
        }
      }
    >
      <div
        className="addressFormWrapper"
      >
        <LoqateAddress
          errorMessage=""
          formName="address"
          manualFields={
            Array [
              Object {
                "label": "Postcode",
                "name": "postCode",
                "onChange": [Function],
                "required": true,
                "status": "default",
                "step": 0,
                "validationType": "postCode",
                "value": "",
                "warning": "* Please type correct UK postal code",
              },
              Object {
                "label": "House name",
                "name": "houseName",
                "onChange": [Function],
                "status": "default",
                "step": 0,
                "validationType": "none",
                "value": "",
              },
              Object {
                "label": "House number",
                "name": "houseNumber",
                "onChange": [Function],
                "status": "default",
                "step": 0,
                "validationType": "none",
                "value": "",
              },
              Object {
                "label": "Street",
                "name": "street",
                "onChange": [Function],
                "required": true,
                "status": "default",
                "step": 0,
                "validationType": "notEmpty",
                "value": "",
              },
              Object {
                "label": "District",
                "name": "district",
                "onChange": [Function],
                "status": "default",
                "step": 0,
                "validationType": "none",
                "value": "",
              },
              Object {
                "label": "Town",
                "name": "town",
                "onChange": [Function],
                "required": true,
                "status": "default",
                "step": 0,
                "validationType": "notEmpty",
                "value": "",
              },
              Object {
                "label": "County",
                "name": "county",
                "onChange": [Function],
                "status": "default",
                "step": 0,
                "validationType": "none",
                "value": "",
              },
            ]
          }
          selectHandler={[Function]}
          subTitle="Start typing your address and select the relevant option when it appears in the list below. If it is not found then click on the 'Enter address manually' link"
        >
          <Spinner
            className=""
          >
            <div
              className="container"
            >
              <div
                className="loader"
              />
            </div>
          </Spinner>
        </LoqateAddress>
      </div>
    </PostCodeStep>
    <InputCheckbox
      checked={false}
      className=""
      customClasses="inlineFlex"
      data-id="I confirm the vehicle will be kept at this address"
      label="I confirm the vehicle will be kept at this address"
      name="I confirm the vehicle will be kept at this address"
      onChange={[Function]}
      status=""
    >
      <div
        className="wrapper inlineFlex"
      >
        <input
          checked={false}
          className=""
          data-id="I confirm the vehicle will be kept at this address"
          name="I confirm the vehicle will be kept at this address"
          onChange={[Function]}
          type="checkbox"
        />
        <Label
          customClasses="inlineFlex"
          name="I confirm the vehicle will be kept at this address"
          status=""
          title="I confirm the vehicle will be kept at this address"
          withSub={false}
        >
          <div
            className="label inlineFlex"
          >
            <label
              data-id={false}
              htmlFor="I confirm the vehicle will be kept at this address"
            >
              <span
                className="title"
                data-id="label-title"
              >
                I confirm the vehicle will be kept at this address
              </span>
            </label>
          </div>
        </Label>
      </div>
    </InputCheckbox>
    <Button
      className="confirm"
      disabled={false}
      href=""
      id="ssFinanceHomeAddressSubmit"
      onClick={[Function]}
      selected={false}
      styleType="primary"
      type="submit"
    >
      <button
        className="buttonComponent primary confirm"
        disabled={false}
        href={null}
        id="ssFinanceHomeAddressSubmit"
        onClick={[Function]}
        type="submit"
      >
        Next
      </button>
    </Button>
  </div>
</Address>
`;
